src/main/java/de/thomas_oster/liblasercut/drivers/ThunderLaser.java

```diff
@@ -0,0 +1,1020 @@
+            + focus2mm(MINFOCUS) + "mm to " + focus2mm(MAXFOCUS) + "mm.");
+        }
+      }
+      if (p instanceof Raster3dPart)
+      {
+        Raster3dPart rp = (Raster3dPart) p;
+        float focus = rp.getLaserProperty() == null ? 0 : ((ThunderLaserProperty)rp.getLaserProperty()).getFocus();
+        if (mm2focus(focus) > MAXFOCUS || (mm2focus(focus)) < MINFOCUS)
+        {
+          throw new IllegalJobException("Illegal Focus value. This Lasercutter supports values between"
+            + focus2mm(MINFOCUS) + "mm to " + focus2mm(MAXFOCUS) + "mm.");
+        }
+      }
+    }
+  }
+
+  /**
+   * It is called, whenever VisiCut wants the driver to send a job to the lasercutter.
+   * @param job This is an LaserJob object, containing all information on the job, which is to be sent
+   * @param pl Use this object to inform VisiCut about the progress of your sending action.
+   * @param warnings If you there are warnings for the user, you can add them to this list, so they can be displayed by VisiCut
+   * @throws IllegalJobException Throw this exception, when the job is not suitable for the current machine
+   * @throws Exception
+   */
+  @Override
+  public void sendJob(LaserJob job, ProgressListener pl, List<String> warnings) throws IllegalJobException, Exception
+  {
+    System.out.println("JOB title >" + job.getTitle() + "< name >" + job.getName() + "< user >"+ job.getUser() + "<");
+
+//    pl.progressChanged(this, 0);
+//    pl.taskChanged(this, "checking job");
+    checkJob(job);
+    job.applyStartPoint();
+
+    Ruida ruida = new Ruida();
+    ruida.setName(job.getTitle());
+
+    for (JobPart p : job.getParts())
+    {
+      float focus;
+
+      if (p instanceof Raster3dPart || p instanceof RasterPart)
+      {
+        RasterPart rp = (RasterPart)p;
+        double dpi = rp.getDPI();
+        Point sp = rp.getRasterStart();
+        int width = rp.getRasterWidth();
+        int height = rp.getRasterHeight();
+        System.out.println(String.format("RasterPart(%d x %d pixels) @ %.1f dpi", width, height, dpi));
+        double rwidth = Util.px2mm(width, dpi);
+        double rheight = Util.px2mm(height, dpi);
+
+        // ruida x,y in mm
+        double rx = Util.px2mm(sp.x, dpi);
+        double ry = Util.px2mm(sp.y, dpi);
+        System.out.println(String.format("RasterPart(%.4fm, %.4f mm) - (%.4f, %.4f mm): %sdirectional", rx, ry, rwidth, rheight, (this.useBidirectionalRastering)?"bi":"uni"));
+        ruida.startPart(rx, ry, rwidth, rheight);
+
+        ThunderLaserProperty prop = (ThunderLaserProperty) rp.getLaserProperty();
+        ruida.setMinPower((int)prop.getMinPower());
+        ruida.setMaxPower((int)prop.getPower());
+        ruida.setSpeed((int)prop.getSpeed());
+        ruida.setFrequency((int)prop.getFrequency());
+        // focus ?
+
+        boolean leftToRight = true; // start with left-to-right
+        for (int y = 0; y < height; y++) { // height
```

Comment:
Thanks, need to test ...

Reply:
just as a remark: you may (maybe, just guessing) need to change your "layer detection" so that it doesn't detect engrave3d as billions of "layers", because it changes power on every grayscale change. (not sure if engrave3d is even supported by the current code)

Reply:
Using `convertRasterizableToVectorPart` now